# å…¨åŸºå› ç»„å¤åˆ¶åçš„äº’è¡¥ä¸¢å¤±

# 1. ç­›é€‰åŸºå› å®¶æ—

å› ä¸ºç›®çš„æ˜¯çœ‹å¤–ç¾¤ä¸­ä¿ç•™ä¸‹æ¥äº†wgdçš„ä¸¤ä»½æ‹·è´ï¼Œè€Œç›®æ ‡çš„ç‰©ç§å¯¹ä¸­åˆ†åˆ«ä¿ç•™äº†ä¸åŒçš„æ—ç³»åŒæºåŸºå› ï¼Œæ‰€ä»¥å¯ä»¥å…ˆæ ¹æ®åŸºå› å®¶æ—ä¸­ç›®æ ‡ç‰©ç§çš„åŸºå› æ•°ç›®è¿›è¡Œç­›é€‰ï¼Œè¦æ±‚ï¼š1.å¤–ç¾¤è‡³å°‘æœ‰ä¸€ä¸ªåŸºå› ï¼›2.è¿‘ç¼˜ç§æœ‰è‡³å°‘ä¸¤ä¸ªåŸºå› ï¼›3.ç›®æ ‡ç‰©ç§å¯¹å„è‡³å°‘æœ‰ä¸€ä¸ªåŸºå› ã€‚å…¶æ¬¡ä¸ºäº†æ–¹ä¾¿åç»­åˆ†æï¼Œå°†`Orthogoups.txt`æ–‡ä»¶ä¸­çš„å…¶ä»–ç‰©ç§åŸºå› åºåˆ—åˆ é™¤ï¼Œåªä¿ç•™æ„Ÿå…´è¶£çš„åŸºå› åˆ—è¡¨å³å¯ã€‚æœ€ç»ˆè·å¾—ä¸¤ä¸ªæ–‡ä»¶ï¼š

`FilteredOrthogroups.txt`æ–‡ä»¶å°±ç®€å•ç²—æš´çš„æŠŠåˆ«çš„ç‰©ç§çš„åŸºå› åˆ æ‰å°±è¡Œäº†ï¼Œä¸ºäº†ç”»æ ‘ä¹‹åæ–¹ä¾¿å®šæ ¹ï¼Œå»ºè®®é‡‡ç”¨4ä¸ªç‰©ç§ï¼š1.å¤–ç¾¤ï¼›2.å…±äº«wgdçš„è¿‘ç¼˜ç§ï¼›3.ä¸¤ä¸ªç›®æ ‡ç‰©ç§ã€‚

- FilteredOrthogroups.list
  
    ```
    OG0000000
    OG0000017
    ...
    OG0001117
    ```
    
- FilteredOrthogroups.txt
  
    ```
    OG0000085:      Bcy|TJE003394   Bcy|TJE007772    Bgy|evm.model.Bruguiera_gymnorrhiza_PacBio_hic_scaffold_16.428
    ```
    

---

# 2. æå–åºåˆ—

ğŸ’¡ è®°å¾—æ ¹æ®æå–çš„æ˜¯è›‹ç™½è´¨è¿˜æ˜¯cdsåºåˆ—ä¿®æ”¹è„šæœ¬38è¡Œçš„åç¼€ã€‚
**ä¸€å®šè¦æ³¨æ„è›‹ç™½è´¨åºåˆ—å’Œcdsåºåˆ—å‘½åæ˜¯å¦ä¸€è‡´ï¼Œå°¤å…¶æ˜¯phytozomeä¸Šé¢ä¸‹è½½çš„æ–‡ä»¶å¸¸å¸¸åŸºå› ååé¢æœ‰ä¸ª.pä¹‹ç±»çš„ï¼Œåˆ‡è®°å»é™¤ï¼ï¼**

```perl
perl 01Ortho2Fa.pl FilteredOrthogroups.list FilteredOrthogroups.txt [Seq file]
mkdir pepdir
mkdir cdsdir
mv *pep pepdir
mv *cds cdsdir
```

**01Ortho2Fa.pl**

```perl
### This script is used to extract the corresponding protein or cds sequences of genes in the genefamilies.
### Change the postfix of output files in LINE 38 according to the type of sequences.

use 5.010;
use autodie;
use Bio::SeqIO;

say "Usage: perl $0 [Target Orthogroup List] [Orthogroups.txt] [Protein File]" if @ARGV<3;

open TARGETORTHO,"<$ARGV[0]";
open ORTHO,"<$ARGV[1]";
my $pepfile=$ARGV[2];

my %pep;
my $fa=Bio::SeqIO->new(-file=>$pepfile,-format=>'fasta');
while (my $seq_obj=$fa->next_seq){
    my $pep_name=$seq_obj->display_name;
    my $seq=$seq_obj->seq;
    my ($spe,$gene)=split/\|/,$pep_name;
    $pep_name="$gene"."_"."$spe";
    $pep{$pep_name}=$seq;
}

my @TargetOrtho;
while (<TARGETORTHO>) {
        s/\s+$//;
        my $orth=(split/\s+/,$_)[0];
        push @TargetOrtho,$orth;
}
close TARGETORTHO;

while (<ORTHO>) {
        s/\s+$//;
        my @eles=split/\s+/;
        my $ortho=shift @eles;
        $ortho=~s/://;
        if ($ortho~~@TargetOrtho) {
                open OUT,">$ortho.cds";
                map {
                        my ($spe,$gene)=split/\|/,$_;
                my $pep_name="$gene"."_"."$spe";
                print OUT ">$pep_name\n$pep{$pep_name}\n";
                } @eles;
                close OUT;
        }
}
close ORTHO;
```

---

# 3. mafftè¿›è¡Œè›‹ç™½è´¨æ¯”å¯¹

ğŸ’¡ å¦‚æœmafftæ–‡ä»¶è¡Œæ•°è¾ƒå¤šï¼Œå¯ä»¥é€šè¿‡splitå‘½ä»¤è¿›è¡Œåˆ†å‰²ï¼Œç„¶åå¹¶è¡Œã€‚
`split -l 1000 [mafft.sh]`

```perl
perl 02CreateMafftSh.pl pepdir mafft.sh mafftoutdir
bash mafft.sh
```

**02CreateMafftSh.pl**

```perl
### Add the path to Mafft software in Line 16 if Mafft is not added to environment variable.

use warnings;
use strict;
my $pepdir=shift;
my @files=<$pepdir/*pep>;
my $output_sh=shift;
open (O,">$output_sh");

my $outdir=shift;
system("mkdir $outdir");
foreach my $file (@files){
    $file=~/OG(\d+)/;
    my $sign=$1;
    my $outfile="$outdir/mafft$1";
    print O "mafft --maxiterate 1000 --localpair $file >$outfile\n";
}
close O;
```

---

# 3. æ ¹æ®è›‹ç™½è´¨åºåˆ—æå–cdsåºåˆ—

```bash
perl 03CreatePal2nalSh.pl mafftoutdir cdsdir pal2naloutdir pal2nal.sh
## åŒæ ·çš„ï¼Œå¦‚æœè¡Œæ•°æ¯”è¾ƒå¤šï¼Œå¯ä»¥ç”¨splitçš„æ–¹æ³•åˆ†å‰²ï¼Œç„¶åå¹¶è¡Œã€‚
```

**03CreatePal2nalSh.pl**

```perl
use strict;
my $pepdir=shift;
my @mafftfiles=<$pepdir/*>;
my $cdsdir=shift;
my $outdir=shift;
system ("mkdir $outdir");
my $output_sh=shift;
open(O,">$output_sh");

my $pal2alnpath="/public1/users/lisen/software/pal2nal.v14/pal2nal.pl";
foreach my $pepfile (@mafftfiles){
    $pepfile=~/mafft(\d+)/;
    my $sign=$1;
    my $cdsfile="OG$sign".".cds";
    print O "perl $pal2alnpath $pepfile $cdsdir/$cdsfile -output fasta >$outdir/pal2nal_$sign\n";
}
close O;
```

---

# 4. raxml-ngå»ºæ ‘

## 4.1 å°†faè½¬æ¢ä¸ºphyæ ¼å¼

```bash
cd pal2naloutdir
perl 04Fa2Phy.pl
mkdir phys fas
mv *phy phys
mv pal* fas
```

**04Fa2Phy.pl**

```perl
use Bio::SeqIO;

my @files=glob("pal*");

foreach my $file (@files) {
        my $out="$file.phy";
        my %seq;
        my $seq_num=0;
        my $seq_length;
        my $fa=Bio::SeqIO->new(-file=>$file,-format=>'fasta');
        while (my $obj=$fa->next_seq) {
        my $seq_name=$obj->display_name;
        my $seq=$obj->seq;
        $seq{$seq_name}=$seq;
        $seq_num++;
        $seq_length=length $seq;
        }
        open OUT,">$out";
        print OUT "$seq_num $seq_length\n";
        foreach my $seq_name (keys %seq) {
                print OUT "$seq_name $seq{$seq_name}\n";
        }
        close OUT;
}
```

## 4.2 raxml-ngå»ºæ ‘

å»ºè®®ç”¨v1.1ï¼Œv0.9æœ‰æ—¶å€™ä¼šæŠ¥é”™ã€‚56æœåŠ¡å™¨æ˜¯å¯ä»¥ç”¨v1.1çš„ï¼Œè€Œ41ä¸å¯ä»¥ã€‚å¦‚æœå‡ºç°likelihoodå¤ªä½çš„æŠ¥é”™ï¼Œå¯ä»¥è¯•ä¸€ä¸‹GTR+G modelã€‚

**05CreateRaxmlngSh.pl**

```perl
use warnings;
use strict;

my @files=glob("*phy");

L: foreach my $file (@files) {
	(my $prefix=$file)=~s/\.phy//;
	my $err="$prefix.err";
	print "raxml-ng --all -msa $file --model GTR+I+G --prefix $prefix --seed 2 --threads 1 --bs-metric fbp,tbe --bs-trees 100 2>>$err\n";
}
```

```perl
cd phys
perl 05CreateRaxmlngSh.pl >raxml.sh
nohup bash raxml.sh &
## åŒæ ·çš„ï¼Œå¦‚æœè¡Œæ•°æ¯”è¾ƒå¤šï¼Œå¯ä»¥ç”¨splitçš„æ–¹æ³•åˆ†å‰²ï¼Œç„¶åå¹¶è¡Œã€‚
```

---

# 5. notungåšre-root

```bash
echo ../species.tree >batch.txt
ls *bestTree >>batch.txt
nohup java -jar -Xmx8g -XX:ParallelGCThreads=4 ~/software/Notung-2.9.1.5.jar -b batch_tree.txt --root --speciestag postfix --edgeweights name --treeoutput nhx --nolosses --outputdir ../root/
```

---

# 6. ç»Ÿè®¡ä¸²è”é‡å¤

è¿™é‡Œåªå…³å¿ƒæ¯ä¸ªåŸºå› å®¶æ—ä¸­ç›®æ ‡çš„ä¸‰ä¸ªç‰©ç§çš„ä¸²è”å¤åˆ¶æƒ…å†µï¼Œå¯ä»¥ä»gffæ–‡ä»¶è¿›è¡Œç­›é€‰ï¼Œè¦æ±‚å¦‚æœä¸¤ä¸ªåŸºå› ä¹‹é—´ç›¸éš”çš„åŸºå› ä¸è¶…è¿‡5ä¸ªå°±è®¤ä¸ºæ˜¯ä¸²è”å¤åˆ¶ã€‚